k2/csrc/properties.cc

```diff
@@ -0,0 +1,31 @@
+#include "fsa.h"
+#include "properties.h"
+
+namespace k2{
+    
+    bool IsTopSorted(const Fsa &fsa){
+        for(auto &range: fsa.leaving_arcs){
+            for(std::size_t arc_idx = range.begin; arc_idx < range.end; ++arc_idx){
+                const Arc &arc = fsa.arcs[arc_idx];
+                if(arc.dest_state <= arc.src_state){
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    bool HasSelfLoops(const Fsa &fsa){
+        //TODO: refactor code below as we have so many for-for-loop structures 
```

Comment:
Incidentally, the range thing may not be the best way of doing that as they'd always be consecutive, it might be better to have a vector of size num_states to store the first leaving_arc of each state.  (Note: loops don't have to include the very last state which is always the final-state which has no arcs leaving it, so safe to use that_vec[state_index+1] as the end).

Reply:
Ok, let me think about this.

Reply:
To reduce the number of loops, we can iterate through only the arcs:

```.cpp
for (const auto& arc: fsa.arcs) {
  if (arc.dest_state == arc.src_state) {
    return true; 
  }
} 
return false;
```

The same applies to `IsTopSorted()`. There is no need to iterate through
all the states.

-----

I guess it is a bug  for the following `if` statement
https://github.com/danpovey/k2/blob/44686b6888b0c7921ef4570f03094fe71a35aaee/k2/csrc/properties.cc#L31-L32

It should return `true` when `src_state == dst_state`, which indicates there
is a self loop; so `HasSelfLoops` should return `true`, not `false`.